### PR TITLE
api: silence warning if API response failed due to connection closed

### LINKED ISSUE #36109: CI: Cilium E2E Upgrade - no-errors-in-logs - Cilium API handler panicked
## CI failure

- Hit three times in https://github.com/cilium/cilium/actions/runs/11958841576
- Sysdump: [cilium-sysdump-6-final.zip](https://github.com/user-attachments/files/17866835/cilium-sysdump-6-final.zip)

```
❌ Found 1 logs in kind-kind/kube-system/cilium-n8nvt (cilium-agent) matching list of errors that must be investigated:
time="2024-11-21T18:06:53.413851782Z" level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" subsys=api url=/v1/healthz (1 occurrences)
```

### PR DESCRIPTION
**Pull request by giorio94**
The Cilium API handlers are wrapped by a logic that takes care of recovering from potential panics triggered by the handlers themselves, emitting an associated warning. One such panic occurs when it is not possible to write the response to the client, due to the connection having already been terminated (as WriteResponse just panics on error). Considering that this condition is expected (e.g., the client may have timed out in the meanwhile \[1]), let's slightly adapt the wrapper to lower the severity of this specific error to debug level, and avoid printing the associated stack trace.

\[1]: for instance, it is possible to trigger this condition by running any cillium-dbg command that invokes an API in a loop, configuring a very short timeout.

Fixes: https://github.com/cilium/cilium/issues/36109

### CODE REVIEW

File: pkg/api/apipanic.go [hunk 1]
```diff
@@ -24,12 +26,17 @@ func (h *APIPanicHandler) ServeHTTP(wr http.ResponseWriter, req *http.Request) {
...
+				"client": req.RemoteAddr,
 			}
-			log.WithFields(fields).Warn("Cilium API handler panicked")
+
+			if err, ok := r.(error); ok && errors.Is(err, syscall.EPIPE) {
+				log.WithError(err).WithFields(fields).Debug("Failed to write API response")
+				return
+			}
+
+			log.WithFields(fields).WithField("panic_message", r).Warn("Cilium API handler panicked")
```
Comment (learnitall) [inline pkg/api/apipanic.go:39]:
```suggestion
			log.WithFields(fields).WithField("panic_message", r).Warn("Cilium API handler panicked in an unexpected way")
```

File: pkg/api/apipanic.go [hunk 2]
```diff
@@ -24,12 +26,17 @@ func (h *APIPanicHandler) ServeHTTP(wr http.ResponseWriter, req *http.Request) {
...
-				"method":        req.Method,
-				"client":        req.RemoteAddr,
+				"url":    req.URL.String(),
+				"method": req.Method,
+				"client": req.RemoteAddr,
 			}
-			log.WithFields(fields).Warn("Cilium API handler panicked")
+
+			if err, ok := r.(error); ok && errors.Is(err, syscall.EPIPE) {
+				log.WithError(err).WithFields(fields).Debug("Failed to write API response")
```
Comment (learnitall) [inline pkg/api/apipanic.go]:
```suggestion
				log.WithError(err).WithFields(fields).Debug("Failed to write API response: client connection closed")
```

**Review by learnitall — COMMENTED** [review]
Nice! Left a couple of small suggestions for helping to differentiate the two error cases. I think the main important bit here is to make sure the debug log on line 35 is different from the debug log on line 45. WDYT?

**Comment by giorio94** [comment]
> I think the main important bit here is to make sure the debug log on line 35 is different from the debug log on line 45. WDYT?

I had previously made the messages identical because they could still be distinguished looking at the associated fields, which are different, but yeah, your suggestion makes sense to make the difference more explicit.

File: pkg/api/apipanic.go (hunk 1 above)
Comment (giorio94) [reply pkg/api/apipanic.go:39]: Hmm, to me the term panic already conveys that it is an unexpected event (I haven't actually introduced/changed this log message, just moved the `panic_message` field after removing it from the common ones). Happy to change the message if you feel the other clearer though.

Comment (learnitall) [reply pkg/api/apipanic.go:39]: Ah that's a good point, we can leave it as is.

**Review by learnitall — APPROVED** [review]
Thank you!

**Pull request merged**